apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/EphemeralTransaction.cs

```diff
@@ -0,0 +1,74 @@
+{
+    /// <summary>
+    ///     EphemeralTransaction acquires sessions from the provided SpannerConnection on an as-needed basis.
+    ///     It holds no long term refcounts or state of its own and can be created and GC'd at will.
+    /// </summary>
+    internal sealed class EphemeralTransaction : ISpannerTransaction
+    {
+        private readonly SpannerConnection _connection;
+
+        public EphemeralTransaction(SpannerConnection connection)
+        {
+            connection.AssertNotNull(nameof(connection));
+            _connection = connection;
+        }
+
+        /// <summary>
+        ///     Acquires a read/write transaction from Spannerconnection and releases the transaction back into the pool
+        ///     after the operation is complete.
+        ///     SpannerCommand never uses implicit Transactions for write operations because they are non idempotent and
+        ///     may result in unexpected errors if retry is used.
+        /// </summary>
+        /// <param name="mutations"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public Task<int> ExecuteMutationsAsync(List<Mutation> mutations, CancellationToken cancellationToken)
+        {
+            mutations.AssertNotNull(nameof(mutations));
+            Logger.Debug(() => "Executing a mutation change through an ephemeral transaction.");
+            int count;
+
+            return ExecuteHelper.WithErrorTranslationAndProfiling(async () =>
+            {
+                using (var transaction = await _connection.BeginTransactionAsync(cancellationToken)
+                    .ConfigureAwait(false))
+                {
+                    count = await ((ISpannerTransaction) transaction)
+                        .ExecuteMutationsAsync(mutations, cancellationToken)
+                        .ConfigureAwait(false);
+                    await transaction.CommitAsync().ConfigureAwait(false);
+                }
+                return count;
+            }, "EphemeralTransaction.ExecuteMutations");
+        }
+
+        public Task<ReliableStreamReader> ExecuteQueryAsync(string sql, CancellationToken cancellationToken)
+        {
+            return ExecuteHelper.WithErrorTranslationAndProfiling(async () =>
+            {
+                sql.AssertNotNullOrEmpty(nameof(sql));
+                Logger.Debug(() => "Executing a query through an ephemeral transaction.");
+
+                using (var holder = await SpannerConnection.SessionHolder
+                    .Allocate(_connection, cancellationToken)
+                    .ConfigureAwait(false))
+                {
+                    var streamReader = _connection.SpannerClient.GetSqlStreamReader(new ExecuteSqlRequest {
+                        Sql = sql
+                    }, holder.TakeOwnership());
+
+                    streamReader.StreamClosed += (o, e) => { _connection.ReleaseSession(streamReader.Session); };
+
+                    return streamReader;
```

Comment:
Does disposing of the session not invalidate the stream reader? (This may well be a Spanner subtlety.)